dataRetriever/resolvers/topicResolverSender/topicResolverSender.go

```diff
@@ -237,49 +239,45 @@ func (trs *topicResolverSender) sendOnTopic(
 	return msgSentCounter
 }
 
+func getPeerID(index int, peersList []core.PeerID, preferredPeer core.PeerID, peerType string, topic string, histogramMap map[string]int) core.PeerID {
+	if index == preferredPeerIndex {
+		histogramMap["preferred"]++
+		log.Debug("sending request to preferred peer", "peer", preferredPeer.Pretty(), "topic", topic, "peer type", peerType)
+
+		return preferredPeer
+	}
+
+	histogramMap[peerType]++
+	return peersList[index]
+}
+
 func (trs *topicResolverSender) getPreferredPeer(pType peerType) core.PeerID {
 	peersInShard, found := trs.getPeersBasedOnPeerType(pType)
 	if !found {
 		return ""
 	}
 
-	indexes := make([]int, 0, len(peersInShard))
-	for idx := range peersInShard {
-		indexes = append(indexes, idx)
-	}
-
-	shuffledIndexes := random.FisherYatesShuffle(indexes, trs.randomizer)
+	randomIdx := trs.randomizer.Intn(len(peersInShard))
 
-	return peersInShard[shuffledIndexes[0]]
+	return peersInShard[randomIdx]
 }
 
 func (trs *topicResolverSender) getPeersBasedOnPeerType(pType peerType) ([]core.PeerID, bool) {
 	preferredPeers := trs.preferredPeersHolderHandler.Get()
 
-	var targetShardID uint32
+	var shardToSearch uint32
 	if pType == crossShardPeer {
```

Comment:
this if branch could have been avoided by passing the shard ID directly on the function

Reply:
done